core/src/main/kotlin/io/zeebe/bpmnspec/builder/SpecBuilder.kt

```diff
@@ -0,0 +1,141 @@
+class TestCaseBuilder(private val name: String, private val description: String?) {
+    private val actions = ArrayList<Action>()
+    private val verifications = ArrayList<Verification>()
+
+    fun actions(init: ActionsBuilder.() -> Unit) {
+        val actionsBuilder = ActionsBuilder();
+
+        actionsBuilder.init()
+
+        actions.addAll(actionsBuilder.build())
+    }
+
+    fun verifications(init: VerificationsBuilder.() -> Unit) {
+        val verificationsBuilder = VerificationsBuilder();
+
+        verificationsBuilder.init()
+
+        verifications.addAll(verificationsBuilder.build())
+    }
+
+    fun build() : TestCase {
+        return TestCase(name, description, actions, verifications)
+    }
+}
+
+@TestSpecMarker
+class ActionsBuilder {
+    private val actions = ArrayList<Action>()
+
+    fun awaitElementInstanceState(selector: ElementSelector, state: ElementInstanceState, workflowInstance: String? = null) {
+        actions.add(AwaitElementInstanceStateAction(state, selector.elementId, selector.elementName, workflowInstance))
+    }
+
+    fun cancelInstance(workflowInstance: String? = null) {
+        actions.add(CancelInstanceAction(workflowInstance));
+    }
+
+    fun completeTask(jobType : String, variables: Map<String, Any> = Collections.emptyMap()) {
+        actions.add(CompleteTaskAction(jobType, serializeToJSON(variables)))
+    }
+
+    fun createInstance(bpmnProcessId: String, variables: Map<String, Any> = Collections.emptyMap(), workflowInstanceAlias: String? = null ) {
+        actions.add(CreateInstanceAction(bpmnProcessId, serializeToJSON(variables), workflowInstanceAlias))
+    }
+
+    fun publishMessage(messageName: String, correlationKey: String, variables: Map<String, Object> = Collections.emptyMap()) {
+        actions.add(PublishMessageAction(messageName, correlationKey, serializeToJSON(variables)))
+    }
+
+    fun throwError(jobType: String, errorCode: String, errorMessage: String = "" ) {
+        actions.add(ThrowErrorAction(jobType, errorCode, errorMessage));
+    }
+
+    fun build(): List<Action> {
+        return actions;
+    }
+}
+
+@TestSpecMarker
+class VerificationsBuilder {
+    private val verifications = ArrayList<Verification>()
+
+    fun elementInstanceState(selector: ElementSelector, state: ElementInstanceState, workflowInstance: String? = null) {
+        verifications.add(ElementInstanceStateVerification(state, selector.elementId, selector.elementName, workflowInstance))
+    }
+
+    fun workflowInstanceState(state: WorkflowInstanceState, workflowInstance: String? = null) {
+        verifications.add(WorkflowInstanceStateVerification(state, workflowInstance))
+    }
+
+    fun workflowInstanceVariable(variableName: String, variableValue: String, selector: ElementSelector, workflowInstance: String? = null) {
+        verifications.add(WorkflowInstanceVariableVerification(variableName, variableValue, workflowInstance, selector.elementId, selector.elementName))
+    }
```

Comment:
We should make the selector optional. This would allow to verify variables on any scope.

Same for the following methods, including the incident verification.